syscall_darwin.go

```diff
@@ -18,3 +26,118 @@ func syscall_syscall9X(fn, a1, a2, a3, a4, a5, a6, a7, a8, a9 uintptr) (r1, r2,
 	callc(syscall9XABI0, unsafe.Pointer(&args))
 	return args.r1, args.r2, args.err
 }
+
+// NewCallback converts a Go function to a function pointer conforming to the C calling convention
+// This is useful when interoperating with C code requiring callbacks. The argument is expected to be a
+// function with one uintptr-sized result. The function must not have arguments with size larger than the size
+// of uintptr. Only a limited number of callbacks may be created in a single Go process, and any memory allocated
+// for these callbacks is never released. At least 2000 callbacks can always be created.
+func NewCallback(fn interface{}) uintptr {
+	return compileCallback(fn)
+}
+
+// only increase this if you have added more to the callbackasm function
+const maxCB = 2000 // maximum number of callbacks
+
+var cbs struct {
+	lock  sync.Mutex
+	numFn int                  // the number of functions currently in cbs.funcs
+	funcs [maxCB]reflect.Value // the saved callbacks
+}
+
+type callbackArgs struct {
+	index uintptr
+	// args points to the argument block.
+	//
+	// For cdecl and stdcall, all arguments are on the stack.
+	//
+	// For fastcall, the trampoline spills register arguments to
+	// the reserved spill slots below the stack arguments,
+	// resulting in a layout equivalent to stdcall.
+	//
+	// For arm, the trampoline stores the register arguments just
+	// below the stack arguments, so again we can treat it as one
+	// big stack arguments frame.
+	args unsafe.Pointer
+	// Below are out-args from callbackWrap
+	result uintptr
+}
+
+func compileCallback(fn interface{}) uintptr {
+	val := reflect.ValueOf(fn)
+	if val.Kind() != reflect.Func {
+		panic("type is not a function")
+	}
+	if cbs.numFn >= maxCB {
+		panic("the maximum number of callbacks has been reached")
+	}
+	ty := val.Type()
+	for i := 0; i < ty.NumIn(); i++ {
+		in := ty.In(i)
+		switch in.Kind() {
+		case reflect.Struct, reflect.Float32, reflect.Float64,
+			reflect.Interface, reflect.Func, reflect.Slice,
+			reflect.Chan, reflect.Complex64, reflect.Complex128,
+			reflect.String, reflect.Map, reflect.Invalid:
+			panic("unsupported argument type: " + in.Kind().String())
+		}
+	}
+	if ty.NumOut() > 1 || ty.NumOut() == 1 && ty.Out(0).Size() != ptrSize {
+		panic("callbacks can only have one pointer-sized return")
+	}
+	(&cbs.lock).Lock()
+	defer (&cbs.lock).Unlock()
+	cbs.funcs[cbs.numFn] = val
+	cbs.numFn++
+	return callbackasmAddr(cbs.numFn - 1)
+}
+
+const ptrSize = unsafe.Sizeof((*int)(nil))
+
+const callbackMaxFrame = 64 * ptrSize
+
+// Implemented in zcallback_GOOS_GOARCH.s
+var callbackasmABI0 uintptr
+
+// this exists to remove the unused function error message
+var _ = callbackWrap
```

Comment:
What is "the unused function error message"? I have never seen...

Reply:
It was an IDE warning. There is no issue compiling with or without this line